### ISSUE TITLE
interrupt when import by arangoimp

### ISSUE THREAD

**Issue by name-yy** [arangodb/arangodb#2078]
Hi, when I import data by arangoimp, it will be interrupt and show
`ERROR error message:    Expecting type Array (exception location: /home/abuild/rpmbuild/BUILD/arangodb-3.0.8/arangod/HttpServer/HttpHandler.cpp:157). Please report this error to arangodb.com`

My arangodb cluster:
system: centos7
arangodb version: 3.0.8
cluster: 3 agency, 2 dbserver, 1 coordinator(192.168.1.63:8531)
             cluster was setup by the method of arangodb manual Chapter "Deployment -- Cluster: Processes"

My import command(run this command on one dbserver):
`arangoimp --collection my_coll --batch-size 67108864 --on-duplicate "update" --progress true --file "my_file" --type "csv" --separator "," --log.level "debug" --log.use-local-time true --server.authentication false --server.database my_db --server.endpoint "http+tcp://192.168.1.63:8531"`

context of my imported datafile("my_file"), There are about 6 million lines in the file:

```
"_key","t1","t2","t3","t4"
"11111111","aaa2","bbb2","ccc2",ddd2"
"22222222","aaa2","bbb2","ccc2",ddd2"
```

Error was occurred when the progress 51% and 69%, I tried twice
First try:

```
Starting CSV import...
2016-09-27T15:41:07 [11673] INFO processed 10387456 bytes (3%) of input file
2016-09-27T15:41:07 [11673] INFO processed 20742144 bytes (6%) of input file
2016-09-27T15:41:07 [11673] INFO processed 31096832 bytes (9%) of input file
2016-09-27T15:41:07 [11673] INFO processed 41484288 bytes (12%) of input file
2016-09-27T15:41:07 [11673] INFO processed 51838976 bytes (15%) of input file
2016-09-27T15:41:31 [11673] INFO processed 62193664 bytes (18%) of input file
2016-09-27T15:41:32 [11673] INFO processed 72548352 bytes (21%) of input file
2016-09-27T15:41:32 [11673] INFO processed 82935808 bytes (24%) of input file
2016-09-27T15:41:32 [11673] INFO processed 93290496 bytes (27%) of input file
2016-09-27T15:41:32 [11673] INFO processed 103645184 bytes (30%) of input file
2016-09-27T15:41:32 [11673] INFO processed 113999872 bytes (33%) of input file
2016-09-27T15:42:21 [11673] INFO processed 124387328 bytes (36%) of input file
2016-09-27T15:42:22 [11673] INFO processed 134742016 bytes (39%) of input file
2016-09-27T15:42:22 [11673] INFO processed 145096704 bytes (42%) of input file
2016-09-27T15:42:22 [11673] INFO processed 155451392 bytes (45%) of input file
2016-09-27T15:42:22 [11673] INFO processed 165838848 bytes (48%) of input file
2016-09-27T15:42:22 [11673] INFO processed 176193536 bytes (51%) of input file
2016-09-27T15:43:34 [11673] INFO processed 186548224 bytes (54%) of input file
2016-09-27T15:43:35 [11673] INFO processed 196902912 bytes (57%) of input file
2016-09-27T15:43:35 [11673] INFO processed 207290368 bytes (60%) of input file
2016-09-27T15:43:35 [11673] INFO processed 217645056 bytes (63%) of input file
2016-09-27T15:43:35 [11673] INFO processed 227999744 bytes (66%) of input file
2016-09-27T15:43:35 [11673] INFO processed 238354432 bytes (69%) of input file

2016-09-27T15:44:53 [11673] ERROR error message:    Expecting type Array (exception location: /home/abuild/rpmbuild/BUILD/arangodb-3.0.8/arangod/HttpServer/HttpHandler.cpp:157). Please report this error to arangodb.com
```

second try:

```
Starting CSV import...
2016-09-27T16:00:27 [12118] INFO processed 10387456 bytes (3%) of input file
2016-09-27T16:00:27 [12118] INFO processed 20742144 bytes (6%) of input file
2016-09-27T16:00:27 [12118] INFO processed 31096832 bytes (9%) of input file
2016-09-27T16:00:27 [12118] INFO processed 41484288 bytes (12%) of input file
2016-09-27T16:00:27 [12118] INFO processed 51838976 bytes (15%) of input file
2016-09-27T16:01:02 [12118] INFO processed 62193664 bytes (18%) of input file
2016-09-27T16:01:02 [12118] INFO processed 72548352 bytes (21%) of input file
2016-09-27T16:01:02 [12118] INFO processed 82935808 bytes (24%) of input file
2016-09-27T16:01:02 [12118] INFO processed 93290496 bytes (27%) of input file
2016-09-27T16:01:02 [12118] INFO processed 103645184 bytes (30%) of input file
2016-09-27T16:01:02 [12118] INFO processed 113999872 bytes (33%) of input file
2016-09-27T16:01:31 [12118] INFO processed 124387328 bytes (36%) of input file
2016-09-27T16:01:31 [12118] INFO processed 134742016 bytes (39%) of input file
2016-09-27T16:01:31 [12118] INFO processed 145096704 bytes (42%) of input file
2016-09-27T16:01:31 [12118] INFO processed 155451392 bytes (45%) of input file
2016-09-27T16:01:31 [12118] INFO processed 165838848 bytes (48%) of input file
2016-09-27T16:01:31 [12118] INFO processed 176193536 bytes (51%) of input file

2016-09-27T16:02:43 [12118] ERROR error message:    Expecting type Array (exception location: /home/abuild/rpmbuild/BUILD/arangodb-3.0.8/arangod/HttpServer/HttpHandler.cpp:157). Please report this error to arangodb.com
```

**Comment by saliormoon** [comment]
Maybe the command line statements which you have used are not correct absolutely.

**Comment by name-yy** [comment]
Could you point out what's wrong with statements?
I also get this error when use http request '_api/import'. 
The error is not arise at the beginning, it seems that more data in arangodb, more chance the error arised.

**Comment by dothebart** [comment]
Please note that the batch size is pretty big. If documents composed exceed the maximum server post data size, errors like this may occur. [This stackoverflow question could be related?](http://stackoverflow.com/questions/37751604/bulkimport-arangoimp)

**Comment by name-yy** [comment]
Hi, I'm afraid is not caused by batch size.

### I do some test again:
This time I use the http request '_api/import' with small batch each request, and when error(the same error I described above) occured, I waited for 30 seconds and reimport the data which caused that error, and it successed.
So I continue import by this manner. But when data in arangodb growth bigger, it can't success import anymore. I have halt the import programme for a whole night, and I retry start import on the morning, it failed.

### My cluster setup:
- I use the two-master mode, 3 agencies, 2 master, 2 coordinators
- arangodb verison 3.0.10
- Three agency run on three virtual machines respectively
- There are two physical machines, one db-server and one coordinator run on each machine respectively. And these two physical machines only used for arango, no other programmes run on them.

### collection setup:
- 2 document collections, 2 edge collections
- 4 shardings for each collection
- 2 replica factors for each collection
- all collections's waitForSync=true, and the import option is also set waitForSync=true
- after create the collections, I stop the master_2 process for a few minutes until all the collection leaders transfer to master_1, and then restart master2 process
- when start to import by http request '_api/import', I use the address of coord_1's(on the same machine of master_1)

### There are some phenomena I observed:
1. When the number of documents close to 70 million and number of edges close to 200 million, it will be can't import anymore. I try the whole test twice, it almost the same
2. After I halt the import programme, the two db servers still have massive network I/O and disk I/O(yes, until now, halt for more than 12 hours)
3. When it can't import anymore, the UI of coord_1 also can't open, and UI of coord_2 can open but can't show node infomation etc
4. The size of data file on master_1 is 33G, but the size of data file on master_2 is only 3.1G
5. The free memory and free disk space is still plenty

### Some question:
1. Is it because of there are a lot of data cached in memory other than direct write in disk?
2. Because I set waitForSync=true, the data should be writen to all master nodes before return the http request. So why the size of data file is different on two master nodes?

**Comment by jsteemann** [comment]
@name-yy : a few questions back that may help us a bit in diagnosing what is happening:
* are you importing all data into the document collections first and into the edge collections only after that? Or are you doing an interleaved import that imports fractions into both the document collections and the edge collections? 
* does the import process stop at a specific point, e.g. always when importing into a specific collection, or is this more or less random?
* regarding the different datasizes on the servers: by what attributes are the data sharded (i.e. which `shardKeys` value do you use for the collections)? And how do the values for the shard keys look like? Could you provide some examples?

**Comment by name-yy** [comment]
1. I imports fractions into both the document collections and the edge collections
2. the import process stop randomly at the beginning, but the final stop(I mean after this import error, I can't to import anymore) it seems stopped on importing the edge collection(the biggest collection, about 0.2 billion items)
3. the shard key is "_key" in every collection

- the data in document collection is {"_key":"11111111111", "attr_1":"11111111111", "attr_2": "xx", "attr_3": 0}
- the data in edge collection is {"_key": "1111111111122222222222", "_from": "11111111111", "_to": "22222222222", "attr_1": 23, "attr_2": 1.43, "attr_3": 1477667805}
- in each import, I make a http request to imoprt a batch of 2000~3000 items of documents, and after the request return, I make another http request to import a batch of 3000~4000 items of edges.

**Comment by jsteemann** [comment]
Closing this as duplicate of #2158.
